group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -370,6 +371,155 @@ public int size() {
      * CoordinatorContext holds all the metadata around a coordinator state machine.
      */
     class CoordinatorContext {
+
+        /**
+         * ContextStateMachine is a wrapper on top of the coordinator state machine. The methods
+         * may be invoked by multiple threads, so they are synchronized.
+         */
+        class ContextStateMachine implements CoordinatorPlayback<U> {
```

Comment:
I think that we are pushing the nested classes a bit too far here and it makes it hard to reason about the concurrency for this class because we access attributes outside of it.